storage/remote/write_handler.go

```diff
@@ -75,7 +75,7 @@ func (h *handler) write(ctx context.Context, req *prompb.WriteRequest) (err erro
 	for _, ts := range req.Timeseries {
 		labels := labelProtosToLabels(ts.Labels)
 		for _, s := range ts.Samples {
-			_, err = app.Add(labels, s.Timestamp, s.Value)
+			_, err = app.Append(0, labels, s.Timestamp, s.Value)
 			if err != nil {
 				return err
 			}
```

Comment:
hello, I have a question, why don't reuse 'SeriesRef' returned by 'app.Append'? If the 'SeriesRef' invalid (via compactHead gc), will a new memSeries be created?

Reply:
```go
	for _, ts := range req.Timeseries {
		labels := labelProtosToLabels(ts.Labels)
		// default to 0，need to search `memSeries` by labels at first traverse
		var ref storage.SeriesRef
		// this inner for loop, labels are always the same
		// this `ts` may have multiple samples, `ref` can be reused to avoid searching `memSeries` each traverse
		for _, s := range ts.Samples {
			ref, err = app.Append(ref, labels, s.Timestamp, s.Value)
			if err != nil {
				switch errors.Cause(err) {
				case storage.ErrOutOfOrderSample, storage.ErrOutOfBounds, storage.ErrDuplicateSampleForTimestamp:
					level.Error(h.logger).Log("msg", "Out of order sample from remote write", "err", err.Error(), "series", labels.String(), "timestamp", s.Timestamp)
				}
				return err
			}
		}
```
will there be any problem? multiple samples of a time series have the same labels

Reply:
Oh I see what you mean. 
It’s not typical to send multiple samples for the same timeseries, but your version is more efficient if it did happen.

Reply:
<img width="1503" alt="image" src="https://github.com/prometheus/prometheus/assets/48707349/ea7dd2d2-70f3-4414-83b8-86c1be7d3f1f">

In this scenario, time series data is generated by aggregating logs and written to the prometheus through remote write, so there will be multiple samples;
I grabbed the cpu profile through pprof, and it looks like labelProtosToLabels and seriesHashmap account for about 7.61 percent of the total. If reuse is SeriesRef, it should be much better. Can I submit a pr to optimize this code?

Reply:
https://github.com/prometheus/prometheus/pull/12580

Reply:
> it looks like labelProtosToLabels and seriesHashmap account for about 7.61 percent of the total. If reuse is SeriesRef, it should be much better.

I don't see where `labelProtosToLabels` would change.

I will note that remote-write handling has been optimised in various downstream projects (Cortex, Thanos, Mimir), using many tricks to avoid allocating memory.